framework/src/ics/IntegralPreservingFunctionIC.C

```diff
@@ -0,0 +1,50 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "IntegralPreservingFunctionIC.h"
+#include "Function.h"
+#include "UserObject.h"
+
+registerMooseObject("MooseApp", IntegralPreservingFunctionIC);
+
+defineLegacyParams(IntegralPreservingFunctionIC);
+
+InputParameters
+IntegralPreservingFunctionIC::validParams()
+{
+  InputParameters params = FunctionIC::validParams();
+  params.addRequiredParam<PostprocessorName>(
+      "integral", "Postprocessor providing the integral of the function, for normalization");
+  params.addRequiredParam<Real>("magnitude",
+                                "Desired magnitude of the initial condition upon integration");
+  params.addClassDescription("Function initial condition that preserves an integral");
+  return params;
+}
+
+IntegralPreservingFunctionIC::IntegralPreservingFunctionIC(const InputParameters & parameters)
+  : FunctionIC(parameters),
+    _pp_name(getParam<PostprocessorName>("integral")),
+    _integral(getPostprocessorValue("integral")),
+    _magnitude(getParam<Real>("magnitude"))
+{
+}
+
+Real
+IntegralPreservingFunctionIC::value(const Point & p)
+{
+  const UserObject & pp = _fe_problem.getUserObject<UserObject>(_pp_name);
```

Comment:
Calling this at every point is really bad

Reply:
let s just move that check to initialize()

Reply:
I imagine it s not in the constructor because pps are created after ICs

Reply:
Yes, this is here because I didn't see any method in `InitialCondition` that lets you do an initial setup action outside the constructor (since the postprocessor doesn't exist yet in the constructor). Did I just miss it somewhere?

Reply:
We do call `InitialConditionBase::initialSetup()` from within FEProblemBase before we do the IC loop

Reply:
Huh, I guess Doxygen for a derived class doesn't necessarily show all the things inherited from the base class.. I assumed it would, so I missed that. Fixed it.